### ISSUE TITLE
训练自己的数据

### ISSUE THREAD

**Issue by zhangxue123** [zylo117/Yet-Another-EfficientDet-Pytorch#259]
首先感谢作者提供的代码，我在用最新版本的代码采用D2和D0训练我自己的数据时，tensorboard查看loss, loss最低只能到0.54，500多个epoch也未能收敛，测试时在验证集和训练集精度都基本为0。我的训练集为5000个样本（检测9类目标，标注框有13893个）完全制作为coco格式喂给模型，检查过数据是正确的，请问您知道大概是哪里有问题呢，导致无法检测？

**Comment by zylo117** [comment]
you should check for overfitting.
And 0.54 is not low enough.
When training coco, D0's val cls loss is only 0.25 and val reg loss is only 0.025.

**Comment by Cli98** [comment]
> 首先感谢作者提供的代码，我在用最新版本的代码采用D2和D0训练我自己的数据时，tensorboard查看loss, loss最低只能到0.54，500多个epoch也未能收敛，测试时在验证集和训练集精度都基本为0。我的训练集为5000个样本（检测9类目标，标注框有13893个）完全制作为coco格式喂给模型，检查过数据是正确的，请问您知道大概是哪里有问题呢，导致无法检测？

@zhangxue123 This is quite common in this repo... As far as I know, no customer dataset get satisfying performance by now.

**Comment by zylo117** [comment]
What about shape dataset? It trains well.
This model is very sensitive to batchsize. Not smaller, not larger, it has to be perfect.

**Comment by zhangxue123** [comment]
> What about shape dataset? It trains well.
> This model is very sensitive to batchsize. Not smaller, not larger, it has to be perfect.

I checked the data, no problem, I think the model is probably only good for the COCO dataset because of the neural network search.

**Comment by zhangxue123** [comment]
> @zhangxue123 This is quite common in this repo... As far as I know, no customer dataset get satisfying performance by now.

I checked the data, no problem, I think the model is probably only good for the COCO dataset because of the neural network search.

**Comment by zhangxue123** [comment]
> > 首先感谢作者提供的代码，我在用最新版本的代码采用D2和D0训练我自己的数据时，tensorboard查看loss, loss最低只能到0.54，500多个epoch也未能收敛，测试时在验证集和训练集精度都基本为0。我的训练集为5000个样本（检测9类目标，标注框有13893个）完全制作为coco格式喂给模型，检查过数据是正确的，请问您知道大概是哪里有问题呢，导致无法检测？
> 
> @zhangxue123 This is quite common in this repo... As far as I know, no customer dataset get satisfying performance by now.

 This is quite common in this repo... As far as I know, no customer dataset get satisfying performance by now.

**Comment by JimLee1996** [comment]
One of the critical reasons is that no other vital augmentations is implemented except random flip. LMAO.

**Comment by zylo117** [comment]
Seriously?
No augmentation is applied on official impl except random filp. They have AA, but none is applied by default. If you want it, you can implement it yourself.

**Comment by Cli98** [comment]
> > What about shape dataset? It trains well.
> > This model is very sensitive to batchsize. Not smaller, not larger, it has to be perfect.
> 
> I checked the data, no problem, I think the model is probably only good for the COCO dataset because of the neural network search.

@zhangxue123 Try my code if you want to modify you anchors. 
Here is the link: https://github.com/Cli98/anchor_computation_tool

**Comment by Zheweiqiu** [comment]
> > > What about shape dataset? It trains well.
> > > This model is very sensitive to batchsize. Not smaller, not larger, it has to be perfect.
> > 
> > 
> > I checked the data, no problem, I think the model is probably only good for the COCO dataset because of the neural network search.
> 
> @zhangxue123 Try my code if you want to modify you anchors.
> Here is the link: https://github.com/Cli98/anchor_computation_tool

I tried to re-calculate anchor_scale and anchor_ratio on my own dataset using method in your repo. The result accuracy is 58.11%, what do you think about this result? Is that satisfactory？

**Comment by pierrelzw** [comment]
> This model is very sensitive to batchsize. Not smaller, not larger, it has to be perfect.

Is that mean it's better to use batchsize=8 for D0-D7?  I also train custom data with this repo but the result AP is quite bad. With D3 and D6, i got mAP@0.5=0.4, while i got mAP@0.5=0.85 with CenterNet_res101 using the same dataset.

I didn't change any code but those for logging. My best train and val loss=0.01 with cls_loss =0.82 and reg_loss=0.017. 

After that the model seems overfit: the val_loss is increased to 0.13 and train loss is decreased to 0.05 or less.    

Do you have more suggestions for training the custom data?

**Comment by zylo117** [comment]
effdet is much deeper and has more activations, so it is more easily overfitting. especially when training larger network like d6/d7, AA should be applied like the paper says.
I've trained lots of custom datasets with D0 myself, no extra augmentation is applied, and the results are great. A little bit better than yolov3. I personally recommend D0 to D3, maybe including D4, the others are too painful to train and consumes too much memory.

**Comment by mad-fogs** [comment]
> > 首先感谢作者提供的代码，我在用最新版本的代码采用D2和D0训练我自己的数据时，tensorboard查看loss, loss最低只能到0.54，500多个epoch也未能收敛，测试时在验证集和训练集精度都基本为0。我的训练集为5000个样本（检测9类目标，标注框有13893个）完全制作为coco格式喂给模型，检查过数据是正确的，请问您知道大概是哪里有问题呢，导致无法检测？
> 
> @zhangxue123 This is quite common in this repo... As far as I know, no customer dataset get satisfying performance by now.

Well, i dont it is wise choice to poste words of this kind.
I got prefect performance on my private datasets, at least.

**Comment by zylo117** [comment]
> > > 首先感谢作者提供的代码，我在用最新版本的代码采用D2和D0训练我自己的数据时，tensorboard查看loss, loss最低只能到0.54，500多个epoch也未能收敛，测试时在验证集和训练集精度都基本为0。我的训练集为5000个样本（检测9类目标，标注框有13893个）完全制作为coco格式喂给模型，检查过数据是正确的，请问您知道大概是哪里有问题呢，导致无法检测？
> > 
> > 
> > @zhangxue123 This is quite common in this repo... As far as I know, no customer dataset get satisfying performance by now.
> 
> Well, i dont think it is wise choice to post words of this kind.
> I got prefect performance on my private datasets, at least.

Same here, I actually trained a lot of custom datasets already, all of them are much better than maskrcnn.

**Comment by newjokker** [comment]
> 首先感谢作者提供的代码，我在用最新版本的代码采用D2和D0训练我自己的数据时，tensorboard查看loss, loss最低只能到0.54，500多个epoch也未能收敛，测试时在验证集和训练集精度都基本为0。我的训练集为5000个样本（检测9类目标，标注框有13893个）完全制作为coco格式喂给模型，检查过数据是正确的，请问您知道大概是哪里有问题呢，导致无法检测？

我也遇到了同样的问题，不知道是哪一步出现了错误

**Comment by zylo117** [comment]
> > 首先感谢作者提供的代码，我在用最新版本的代码采用D2和D0训练我自己的数据时，tensorboard查看loss, loss最低只能到0.54，500多个epoch也未能收敛，测试时在验证集和训练集精度都基本为0。我的训练集为5000个样本（检测9类目标，标注框有13893个）完全制作为coco格式喂给模型，检查过数据是正确的，请问您知道大概是哪里有问题呢，导致无法检测？
> 
> 我也遇到了同样的问题，不知道是哪一步出现了错误

underfitting
